components/scream/src/physics/p3/tests/p3_ice_collection_unit_tests.cpp

```diff
@@ -0,0 +1,315 @@
+#include "catch2/catch.hpp"
+
+#include "share/scream_types.hpp"
+#include "share/util/scream_utils.hpp"
+#include "share/scream_kokkos.hpp"
+#include "share/scream_pack.hpp"
+#include "physics/p3/p3_functions.hpp"
+#include "physics/p3/p3_functions_f90.hpp"
+#include "share/util/scream_kokkos_utils.hpp"
+
+#include "p3_unit_tests_common.hpp"
+
+#include <thread>
+#include <array>
+#include <algorithm>
+#include <random>
+
+namespace scream {
+namespace p3 {
+namespace unit_test {
+
+/*
+ * Unit-tests for p3 ice collection functions.
+ */
+template <typename D>
+struct UnitWrap::UnitTest<D>::TestIceCollection {
+
+  static void run_ice_cldliq_bfb()
+  {
+    // Read in tables
+    view_2d_table vn_table; 
+    view_2d_table vm_table; 
+    view_1d_table mu_r_table; view_dnu_table dnu;
+    Functions::init_kokkos_tables(vn_table, vm_table, mu_r_table, dnu);
+
+    static constexpr Int max_pack_size = 16;
+    REQUIRE(Spack::n <= max_pack_size);
+
+    // Load some lookup inputs, need at least one per pack value
+    IceCldliqCollectionData cldliq[max_pack_size] = {
+      //  rho      temp      rhofaci     f1pr04     qitot      qc           nitot      nc
+      {4.056E-03, 4.02E+01, 8.852E-01, 0.174E+00, 1.221E-14, 5.100E-03, 9.558E+04, 9.952E+05},
+      {6.852E-02, 5.01E+01, 8.852E-01, 0.374E+00, 1.221E-13, 4.100E-03, 9.558E+04, 9.952E+05},
+      {8.852E-02, 6.00E+01, 8.900E-01, 0.123E+00, 1.221E-12, 3.100E-03, 9.558E+04, 9.952E+05},
+      {1.902E-01, 1.00E+02, 8.900E-01, 0.123E+00, 1.221E-11, 2.100E-03, 9.558E+04, 9.952E+05},
+
+      {2.201E-01, 2.00E+02, 0.100E+01, 0.174E+00, 1.221E-10, 1.100E-03, 2.558E+05, 9.952E+06},
+      {3.502E-01, 3.00E+02, 0.100E+01, 0.374E+00, 1.221E-09, 8.100E-04, 2.558E+05, 9.952E+06},
+      {4.852E-01, 5.00E+02, 0.100E+01, 0.123E+00, 1.221E-08, 4.100E-04, 2.558E+05, 9.952E+06},
+      {5.852E-01, 8.00E+02, 0.100E+01, 0.123E+00, 1.221E-07, 2.100E-04, 2.558E+05, 9.952E+06},
+
+      {6.852E-01, 1.00E+03, 0.950E+00, 0.150E+00, 1.221E-06, 9.952E-05, 4.596E+05, 1.734E+07},
+      {7.852E-01, 2.00E+03, 0.950E+00, 0.374E+00, 1.221E-05, 4.952E-05, 4.596E+05, 1.734E+07},
+      {8.852E-01, 4.00E+03, 0.950E+00, 0.123E+00, 1.221E-04, 1.952E-05, 4.596E+05, 1.734E+07},
+      {9.852E-01, 6.00E+03, 0.950E+00, 0.123E+00, 1.221E-03, 9.952E-06, 4.596E+05, 1.734E+07},
+
+      {1.002E+01, 1.00E+04, 1.069E+00, 0.174E+00, 1.221E-02, 6.952E-06, 6.596E+05, 1.734E+08},
+      {1.152E+01, 2.00E+04, 1.069E+00, 0.374E+00, 1.221E-02, 3.952E-06, 6.596E+05, 1.734E+08},
+      {1.252E+01, 4.00E+04, 1.069E+00, 0.123E+00, 1.221E-02, 1.952E-06, 6.596E+05, 1.734E+08},
+      {1.352E+01, 8.00E+04, 1.069E+00, 0.123E+00, 1.221E-02, 9.952E-07, 6.596E+05, 1.734E+08}
+    };
+
+    // Get data from fortran
```

Comment:
I've found it's safer to do the device sync before getting the data from fortran. That way, you don't sync out/in-out data to the device which could alter the results. If the function being tested has no in-out data and doesn't accumulate out data, then it doesn't matter, but I still think it's a good habit to get in to.

In other words, since we are doing a fair bit of copy/pasting syncing in this effort, we want to have the best/safest patterns in place

Reply:
@xyuan , I'd still like you to flip the "Get data from fortran" and the "Sync to device" blocks.